store/database.go

```diff
@@ -72,6 +72,11 @@ type VersionedDatabase interface {
 	VersionedIteratorCreator
 	VersionedBatcher
 
+	// Prune attempts to prune all versions up to and including the provided
+	// version argument. The operation should be idempotent. An error should be
+	// returned upon failure.
+	Prune(version uint64) error
```

Comment:
I'd like to pass the pruning option to the dedicated db, instead of the exposed endpoint.
- avoid the conflicts with sync
- it would be easy to apply pruning at the dedicated backend, (background running)

Reply:
The problem when passing it to the SS engine is _when_ do you prune? The SS interface is just a CRUD API -- there's no notion of "committing".

So when the RootStore commits, how/when does the backing SS store know what/when to prune? You'd need to either do it:

* On every `Put` and `Batch.Write` call
  * Note, doing it on `Batch.Write` would be essentially analogous to doing it on commitment, but the abstraction seems pretty messy to me personally.
* Or, have some sort of external goroutine process that periodically checks and prunes.

So there's tradeoffs to each, but to me having the RootStore simply call `Prune` on the SS backend seems the cleanest.

Thoughts @tac0turtle or @kocubinski?

Reply:
tbh, I have no exact idea how the sync flow looks like but the original store has snapshot and pruning managers and do some duplicated checks like version conflicting. so I'd like to pass this logic to the dedicated backend and maybe can manage it in background routine or whatever the dedicated backend wants

Reply:
That introduces additional complexity and goroutine management, but I can add that to the design doc and we'll discuss next meeting.